hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/LikeFunction.java

```diff
@@ -0,0 +1,214 @@
+        return new LikeFunction(source, pattern, escape);
+    }
+
+    @SuppressFBWarnings(value = "NP_BOOLEAN_RETURN_NULL", justification = "SQL has three-valued boolean logic")
+    @Override
+    public Boolean eval(Row row, ExpressionEvalContext context) {
+        String source = asVarchar(operand1, row, context);
+
+        if (source == null) {
+            return null;
+        }
+
+        String pattern = asVarchar(operand2, row, context);
+
+        if (pattern == null) {
+            return null;
+        }
+
+        String escape;
+
+        if (operand3 != null) {
+            escape = asVarchar(operand3, row, context);
+
+            if (escape == null) {
+                return null;
+            }
+        } else {
+            escape = null;
+        }
+
+        if (state == null) {
+            state = new State();
+        }
+
+        return state.like(source, pattern, escape);
+    }
+
+    @Override
+    public QueryDataType getType() {
+        return QueryDataType.BOOLEAN;
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_LIKE;
+    }
+
+    /**
+     * Helper class to execute LIKE function. Caches the last observed pattern to avoid constant re-compilation.
+     */
+    public static class State {
+        /** Last observed pattern. */
+        private String lastPattern;
+
+        /** Last observed escape. */
+        private String lastEscape;
+
+        /** Last Java pattern. */
+        private Pattern lastJavaPattern;
+
+        public boolean like(String source, String pattern, String escape) {
+            Pattern javaPattern = convertToJavaPattern(pattern, escape);
+
+            Matcher matcher = javaPattern.matcher(source);
+
+            return matcher.matches();
+        }
+
+        private Pattern convertToJavaPattern(String pattern, String escape) {
+            if (lastJavaPattern != null && Objects.equals(pattern, lastPattern) && Objects.equals(escape, lastEscape)) {
+                return lastJavaPattern;
+            }
+
+            String javaPatternStr = constructJavaPatternString(pattern, escape);
+            Pattern javaPattern = Pattern.compile(javaPatternStr);
+
+            lastPattern = pattern;
+            lastEscape = escape;
+            lastJavaPattern = javaPattern;
+
+            return javaPattern;
+        }
+
+        @SuppressWarnings("checkstyle:CyclomaticComplexity")
+        private static String constructJavaPatternString(String pattern, String escape) {
+            // Get the escape character.
+            Character escapeChar;
+
+            if (escape != null) {
+                if (escape.length() != 1) {
+                    throw QueryException.error("Escape parameter should be a single character: " + escape);
```

Comment:
```suggestion
                    throw QueryException.error("ESCAPE parameter must be a single character: " + escape);
```
We can also leak user data to error logs by including the value.

Reply:
Fixed.